### PR TITLE
Add support of utf8mb4 for mysql

### LINKED ISSUE #5660: can't add emoji to issues
- Gitea version (or commit ref): 1.6.3
- Database (use `[x]`):
  - [x] MySQL

## Description
Can't add emoji to issues. I get this error:

`[...outers/repo/issue.go:1192 NewComment()] [E] CreateIssueComment: CreateComment: Error 1366: Incorrect string value: '\xF0\x9F\xA5\xB6' for column 'content' at row 1`

### PR DESCRIPTION
**Pull request by lunny**
This PR will allow you change mysql's charset to `utf8mb4`. This may require your InnoDB greater than 5.6 as  @philfry said on https://github.com/go-gitea/gitea/pull/3516#issuecomment-366648794, but gitea will not check mysql's version currently. Users should know their database version themselves.

should fix #5660, #6988 & #2711

### CODE REVIEW

File: custom/conf/app.ini.sample [hunk 1]
```diff
@@ -260,6 +260,9 @@ PASSWD =
 ; For Postgres, either "disable" (default), "require", or "verify-full"
 ; For MySQL, either "false" (default), "true", or "skip-verify"
 SSL_MODE = disable
+; For MySQL only, either "utf8" or "utf8mb4", default is "utf8".
+; NOTICE: your MySQL InnoDB have to great than 5.6 if this is "utf8mb4" and Gitea will not check that
```
Comment (zeripath) [inline custom/conf/app.ini.sample]:
```suggestion
; NOTICE: for "utf8mb4" you must use MySQL InnoDB > 5.6. Gitea is unable to check this.
```

File: docs/content/doc/advanced/config-cheat-sheet.en-us.md [hunk 2]
```diff
@@ -160,6 +160,7 @@ Values containing `#` or `;` must be quoted using `` ` `` or `"""`.
 - `USER`: **root**: Database username.
 - `PASSWD`: **\<empty\>**: Database user password. Use \`your password\` for quoting if you use special characters in the password.
 - `SSL_MODE`: **disable**: For PostgreSQL and MySQL only.
+- `CHARSET`: **utf8**: For MySQL only, either "utf8" or "utf8mb4", default is "utf8". NOTICE: your MySQL InnoDB have to great than 5.6 if this is "utf8mb4" and Gitea will not check that
```
Comment (zeripath) [inline docs/content/doc/advanced/config-cheat-sheet.en-us.md]:
```suggestion
- `CHARSET`: **utf8**: For MySQL only, either "utf8" or "utf8mb4", default is "utf8". NOTICE: for "utf8mb4" you must use MySQL InnoDB > 5.6. Gitea is unable to check this.
```

File: models/models.go [hunk 3]
```diff
@@ -222,8 +223,8 @@ func getEngine() (*xorm.Engine, error) {
 		if tls == "disable" { // allow (Postgres-inspired) default value to work in MySQL
 			tls = "false"
 		}
-		connStr = fmt.Sprintf("%s:%s@%s(%s)/%s%scharset=utf8&parseTime=true&tls=%s",
-			DbCfg.User, DbCfg.Passwd, connType, DbCfg.Host, DbCfg.Name, Param, tls)
+		connStr = fmt.Sprintf("%s:%s@%s(%s)/%s%scharset=%s&parseTime=true&tls=%s",
+			DbCfg.User, DbCfg.Passwd, connType, DbCfg.Host, DbCfg.Name, Param, DbCfg.Charset, tls)
```
Comment (zeripath) [inline models/models.go:227]: Whilst we're changing this it might be sensible to escape these parts.

File: options/locale/locale_en-US.ini [hunk 4]
```diff
@@ -86,8 +86,9 @@ host = Host
 user = Username
 password = Password
 db_name = Database Name
-db_helper = Note to MySQL users: please use the InnoDB storage engine and the 'utf8_general_ci' character set.
+db_helper = Note to MySQL users: please use the InnoDB storage engine and if you use "utf8mb4", your InnoDB version should great than 5.6 .
```
Comment (zeripath) [inline options/locale/locale_en-US.ini]:
```suggestion
db_helper = Note to MySQL users: please use the InnoDB storage engine and if you use "utf8mb4", your InnoDB version must be greater than 5.6 .
```

**Comment by lunny** [comment]
@zeripath done.

**Review by zeripath — APPROVED** [review]

**Comment by silverwind** [comment]
I take it there is no way to migrate existing `utf8mb3` data because it may already be in a corrupt state?

**Comment by lunny** [comment]
@silverwind migrate from utf8 is possible but it’s out of scope of this PR.

**Review by techknowlogick — APPROVED** [review]

**Comment by techknowlogick** [comment]
Approved, but should we make breaking change and default to utf8mb4?

**Comment by lunny** [comment]
@techknowlogick That could be another PR. Of course, I think utf8mb4 is more reasonable.

**Comment by immanuelfodor** [comment]
Simplest InnoDB version check according to: https://www.fromdual.com/innodb-version

```sql
SHOW GLOBAL VARIABLES LIKE 'innodb_ver%';
```
```
+----------------+-------------+
| Variable_name  | Value       |
+----------------+-------------+
| innodb_version | 5.6.42-84.2 |
+----------------+-------------+
1 row in set (0.00 sec)
```

**Comment by zeripath** [comment]
Ok fancy putting a pr in to test this?

**Comment by mrsdizzie** [comment]
The real thing to test is:

By default there is a  767 byte limit on indexes, which means that you will exceed the index limit length if you have varchar(255) with a 4 byte encoding like utf8mb4 (255 * 4 = 1020). This is fixed by the innodb_large_prefix setting which was introduce in 5.6.3:

 https://dev.mysql.com/doc/refman/5.6/en/innodb-parameters.html#sysvar_innodb_large_prefix

This is still off by default, but many hosts turn it on.

If testing I think the right thing is to test if innodb_large_prefix is set to either 1 or on (both syntax are supported).

```
MariaDB [test]> SHOW GLOBAL VARIABLES LIKE 'innodb_large_prefix%';
+---------------------+-------+
| Variable_name       | Value |
+---------------------+-------+
| innodb_large_prefix | ON    |
+---------------------+-------+
```

More complicated, this feature only works when the ROW_FORMAT is COMPRESSED or DYNAMIC.

The Default for MySQL 5.6 was COMPACT and 5.7 is Dynamic -- so it wouldn't work right on 5.6 unless whatever created the rows also specified ROW_FORMAT=DYNAMIC|COMPRESSED in the create statement (in addition to using innodb_large_prefix) .

To simplify, it might be easiest to require 5.7+ and then check the innodb_large_prefix value to make sure it is 1|ON.

**Comment by immanuelfodor** [comment]
A few weeks ago, I followed the conversion steps of the Nextcloud database as seen here: https://docs.nextcloud.com/server/16/admin_manual/configuration_database/mysql_4byte_support.html Maybe it can be useful for you guys, there were a few SQLs I needed to run, and also an SQL query generator SQL query 😃 FYI, the above InnoDB version you can see in the previous comment is from a MariaDB install on Ubuntu. I don't know how to get 5.7, `apt` installs the 5.6 version by default.

**Comment by silverwind** [comment]
Keep note that the `innodb_large_prefix` was apparently removed in MySQL 8.0. I tried setting it in MySQL 8.0.16 and got this error:

````
unknown variable 'innodb_large_prefix=1'
````

**Pull request merged**